File: design-pattern-jvm/app/src/main/java/design/pattern/java/patterns/behavioral/memento/TextEditorOriginator.java
package design.pattern.java.patterns.behavioral.memento;

class TextEditorOriginator {
  private String state;

  public TextEditorOriginator() {
    this.state = "";
  }

  public TextEditorOriginator(String initialValue) {
    this.state = initialValue == null ? "" : initialValue;
  }

  public TextEditorSnapshot save() {
    return new TextEditorSnapshot(this.state);
  }

  public void restore(TextEditorMemento prev) {
    this.state = prev.getText();
  }

  public TextEditorOriginator write(String value) {
    this.state += value;
    return this;
  }

  public TextEditorOriginator clearAll() {
    this.state = "";
    return this;
  }

  public String getState() {
    return this.state;
  }
}

File: design-pattern-jvm/app/src/main/kotlin/design/pattern/kotlin/patterns/behavioral/memento/TextEditorOriginator.kt
package design.pattern.kotlin.patterns.behavioral.memento

internal class TextEditorOriginator {
    var state: String
        private set

    constructor() {
        state = ""
    }

    constructor(initialValue: String?) {
        state = initialValue ?: ""
    }

    fun save(): TextEditorSnapshot {
        return TextEditorSnapshot(state)
    }

    fun restore(prev: TextEditorMemento) {
        state = prev.text
    }

    fun write(value: String): TextEditorOriginator {
        state += value
        return this
    }

    fun clearAll(): TextEditorOriginator {
        state = ""
        return this
    }
}